### PR TITLE
Allow user-specified underlying type for enums

### LINKED ISSUE #8028: enum: Underlying type
Enumerations in PowerShell appear to be limited to Int32. Ideally enums should allow inheritance-style setting of the underlying type.

Steps to reproduce
------------------

PowerShell does not appear to allow an underlying type to be defined for an enum. All values are coerced to Int32.

The statement below will raise an exception:
```powershell
enum Fruit { Apple = 0x0FFFFFFFFFFFFFFF }
```

Expected behavior
-----------------

An enumeration wiith underlying type set to [Int64] to be created.

```powershell
enum Fruit { Apple = 0x0FFFFFFFFFFFFFFF }
[Fruit].GetEnumUnderlyingType() -eq [Int64]
```

Actual behavior
---------------

An exception is throw when the attempt to cast fails.

```none
Enumerator value is too large for a System.Int.
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.1.0
PSEdition                      Core
GitCommitId                    6.1.0
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by IISResetMe**
## PR Summary
Fix #8028 

This change adds support for specifiying the underlying type for an enum:

```powershell
enum MyEnum : long 
{
  A = 0x0FFFFFFFFFFFFFFF
  B
}
# or
enum MyByte : byte 
{
  A = 0x01
  B = 0x02
  C = 0x03
  D
}
```

TODO:
 - [x] re-introduce range validation (fixed in f417453b99636bdb1fd10b25d3ea554304d72315)
   - [x] ~~currently breaks on negative value literals~~ (fixed in 19b277054a16139678b64f277e5dd95530638add)
 - [x] better parsing errors for invalid underlying types (fixed in fc5a608f3483941f82d1a817cbde1eead3c18a77)
 - [x] ~~value assignment is currently broken from non-int types~~ (fixed in f58eba81d585a74cdef8e22807e9f39710e406ea)
 - [x] negative parsing tests

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [X] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/PSType.cs]: I think we could do the check in parser. Seems it is more useful.

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/PSType.cs]:
```suggestion
                    if (value != 0 && _enumDefinitionAst.Attributes.Any(attr => attr.TypeName.GetReflectionType() == typeof(FlagsAttribute)))
```

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/PSType.cs]: Please use switch statement.

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/PSType.cs]: Please align arguments to one column (under `type.Extent` here).
Below too.

File: src/System.Management.Automation/resources/ParserStrings.resx
Comment (iSazonov) [inline src/System.Management.Automation/resources/ParserStrings.resx]: Maybe replace `integral type` with `underlying type`?

File: test/powershell/Language/Classes/scripting.enums.tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Classes/scripting.enums.tests.ps1]: The same.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Why not static?
And if we moved this to Parser why we did not remove this from PSType?

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (vexx32) [inline src/System.Management.Automation/engine/parser/PSType.cs]: Given that there is a finite number of acceptable types that are permitted here, wouldn't using a type-based switch here outperform reflection for this lookup?

Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/PSType.cs]: I don't think performance is a major concern here - it's retrieved _once_ per `enum` defined.

Comment (vexx32) [reply src/System.Management.Automation/engine/parser/PSType.cs]: Every Little Bit Counts™️  😄 

(But yeah, I guess that makes sense!)

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/PSType.cs]: Our goal is to move as many checks to the parser as possible. These checks are no longer needed in PSTypes.

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/PSType.cs]: I'm sure the _engine_ should get rid of slow patterns even if they are supposedly rarely used.
We could use the pattern (type.GetTypeCode() and TypeCode.Int16):
https://github.com/dotnet/corefx/blob/master/src/System.Linq.Expressions/src/System/Dynamic/Utils/TypeUtils.cs#L43-L57

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/PSType.cs]: Seems we could move the checks to parser too. Perhaps not in the PR.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/Parser.cs]: I think it should be static and we should use consts like TypeCode.Int16 and method type.GetTypeCode().

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Please align all arguments to one column.
Below too (there is many cases).

File: src/System.Management.Automation/resources/ParserStrings.resx
Comment (iSazonov) [inline src/System.Management.Automation/resources/ParserStrings.resx]: integral -> underlying

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/Parser.cs]: Switching to TypeCode constants sounds like an excellent idea, will try this

File: src/System.Management.Automation/resources/ParserStrings.resx
Comment (IISResetMe) [reply src/System.Management.Automation/resources/ParserStrings.resx]: Name of the error string is already updated, and the error text clearly states that it's because of an underlying type constraint

Comment (iSazonov) [reply src/System.Management.Automation/resources/ParserStrings.resx]: There is 'Expected a builtin __integral__ type'.

Comment (IISResetMe) [reply src/System.Management.Automation/resources/ParserStrings.resx]: No, that's quite intentional. "builtin underlying type" would be ambiguous.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/Parser.cs]: Fixed all instances of `ReportError()` and `ReportIncompleteInput()`

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/PSType.cs]: The same request - could we move the type check to parser?

Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/PSType.cs]: I honestly don't think that's a good idea. Injecting dynamic properties (we don't know the underlying type at compile time) into the AST - the AST being publicly exposed - sounds like asking for trouble. And we're emitting parser errors related to type constraints in the TypeDefiner for classes, it's not like this behavior is inconsistent with the existing implementation

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: @IISResetMe Please move all formatting changes in separate PR. Our policy is avoid mixing style and functional changes.

Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/Parser.cs]: Ah, misunderstood your previous comment then

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/PSType.cs]: Enum looks like const. So we can make all checks at parse time.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/Parser.cs]: My request was only for your new or changed code but if you fixed all code please move this to new PR and keep current PR as small as possible.

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (vexx32) [inline src/System.Management.Automation/engine/parser/PSType.cs]: Shouldn't this be value <= maxValue / 2? 🙂 (or is there a catch I'm unaware of where you can't have an enum value that's equal to the MaxValue of the underlying type?)

Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/PSType.cs]: You're right, it should. Just turns out that that doesn't matter here, because the highest cardinal number you can fit into any integral type is never even :)

```powershell
@([byte],[sbyte],[uint16],[int16],[int],[uint32],[long],[uint64]).ForEach({$_::MaxValue})
255
127
65535
32767
2147483647
4294967295
9223372036854775807
18446744073709551615
```

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/PSType.cs]: I know that we _can_, I'm saying we _shouldn't_ - if I apply [@lzybkr's line of reasoning here](https://github.com/IISResetMe/PowerShell/commit/bfd4be766246bdf33a0ffbb027ba66723fe06b49#r31450712), we'll at the very least have to keep the check in place here even if we also check at parse-time - and I'm not sure how useful that is

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/Parser.cs]: Done, will fix the indentation style issues in a separate PR

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/PSType.cs]: Thanks for clarify! I see the point and agree that the design is good.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: It should be `readonly static`.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Please align arguments in the new code.
(Please push your removed style commit in new PR)

File: test/powershell/Language/Classes/scripting.enums.tests.ps1
Comment (iSazonov) [inline test/powershell/Language/Classes/scripting.enums.tests.ps1]: From line 41. This should works:
```powershell
enum V1 {
    A = -4
    B = [int]::MinValue
    C = -1
    D
}
```

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (lzybkr) [reply src/System.Management.Automation/engine/parser/PSType.cs]: Uh, it used to do something different for enums I think, you'll have to use your favorite decompiler on an older version of PowerShell to double check though.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (lzybkr) [reply src/System.Management.Automation/engine/parser/Parser.cs]: `const` would be preferred if it's allowed.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/Parser.cs]: Done. Will submit new PR once this is done (to avoid merge conflicts with modified code paths in `Parser.cs`)

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/PSType.cs]: In this case that isn't so interesting because we're always passing in a `Type` instance (for which every other instance I can find `Type.GetTypeCode()` is also used)

File: test/powershell/Language/Classes/scripting.enums.tests.ps1
Comment (IISResetMe) [reply test/powershell/Language/Classes/scripting.enums.tests.ps1]: Difference is that we're counting _towards_ zero when the previous value is negative, so the test on line 41 doesn't make much sense in relation to `Int.MinValue`. 

That being said, we should accept `Int.MinValue`, so I added a test for that + next consecutive auto-assignment

**Comment by vexx32** [comment]
@IISResetMe for future reference when you look at adding that `[Flags()]` stuff back in, I think you could probably safely use bit shifts rather than `*= 2` 😉

File: src/System.Management.Automation/engine/parser/PSType.cs
Comment (iSazonov) [reply src/System.Management.Automation/engine/parser/PSType.cs]: I thought it is for future enhancement but if it was removed we can ignore this.

@IISResetMe Please don't set "Resolve conversation" until we finish the code review - it is difficult to understand that comments needs an attention. Please leave this to comment's author or maintainers.

Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/PSType.cs]: @iSazonov got it, sorry about that

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (IISResetMe) [reply src/System.Management.Automation/engine/parser/Parser.cs]: @iSazonov the type constraint for enums is defined as part of the Common Language Specification in ECMA-335 and hasn't changed since 2001 - I think it's unlikely we'll ever have to update it

Comment (lzybkr) [reply src/System.Management.Automation/engine/parser/Parser.cs]: Moreover - it's a public api, the values can't change. This is why we also specify enum values in PowerShell in places where it might not seem necessary, e.g. `TokenKind`.

File: test/powershell/Language/Classes/scripting.enums.tests.ps1
Comment (TravisEz13) [inline test/powershell/Language/Classes/scripting.enums.tests.ps1]: Can we add cases for value to small and too large?

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

**Comment by TravisEz13** [comment]
Can we get committee review for the new syntax?

File: test/powershell/Language/Classes/scripting.enums.tests.ps1
Comment (TravisEz13) [reply test/powershell/Language/Classes/scripting.enums.tests.ps1]: I'm fine with one check for the error case.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and is fine with the syntax

File: test/powershell/Language/Classes/scripting.enums.tests.ps1
Comment (IISResetMe) [reply test/powershell/Language/Classes/scripting.enums.tests.ps1]: @TravisEz13 Do we want to update `ParserStrings.EnumeratorValueTooLarge` is "either too large _or_ too small for a [typename]". 

Currently (since 5.0) we throw:

```powershell
PS C:\>enum E { A = [long]::MinValue }
At line:1 char:13
+ enum E { A = [long]::MinValue }
+              ~~~~~~~~~~~~~~~~
Enumerator value is too large for a System.Int.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : EnumeratorValueTooLarge
```

**Comment by adityapatwardhan** [comment]
@TravisEz13 Please update your review.

File: test/powershell/Language/Classes/scripting.enums.tests.ps1
Comment (TravisEz13) [reply test/powershell/Language/Classes/scripting.enums.tests.ps1]: Sure

**Comment by vexx32** [comment]
@TravisEz13 @IISResetMe apart from the merge conflict, was there anything else pending here? 💖

**Comment by TravisEz13** [comment]
@vexx32 	My comments have not been addressed. https://github.com/PowerShell/PowerShell/pull/8329#pullrequestreview-180404781

**Comment by iSazonov** [comment]
@IISResetMe Please resolve merge conflicts.

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Const should be upper case:
```suggestion
            const TypeCode ValidUnderlyingTypeCodes = TypeCode.Byte | TypeCode.Int16 | TypeCode.Int32 | TypeCode.Int64 | TypeCode.SByte | TypeCode.UInt16 | TypeCode.UInt32 | TypeCode.UInt64;
```

File: src/System.Management.Automation/engine/parser/Parser.cs
Comment (iSazonov) [inline src/System.Management.Automation/engine/parser/Parser.cs]: Please add space:
```suggestion
                    // No need to report error since there is error reported in method StatementRule.
```

**Comment by IISResetMe** [comment]
@iSazonov done!

**Comment by IISResetMe** [comment]
[The Codacy issue currently failing](https://app.codacy.com/app/slee/PowerShell/file/28451399091/issues/source?bid=10638692&fileBranchId=10639194#l1172) is not really relevant.

While the compiler can't prove it, that path is unreachable since `case TypeCode.Int32` will always be chosen if `typeConstraintAst == null`

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**
